File: babbleapp/src/main/java/com/learning/leap/beginningtobabbleapp/settings/UserInfoActivity.java
package com.learning.leap.beginningtobabbleapp.settings;

import android.app.ProgressDialog;
import android.content.Intent;
import android.provider.Settings;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.learning.leap.beginningtobabbleapp.download.DownloadActivity;
import com.learning.leap.beginningtobabbleapp.R;
import com.learning.leap.beginningtobabbleapp.userInfo.UserInfoPresenter;
import com.learning.leap.beginningtobabbleapp.userInfo.UserInfoViewInterface;
import com.learning.leap.beginningtobabbleapp.utility.Utility;
import com.learning.leap.beginningtobabbleapp.models.BabblePlayer;
import com.learning.leap.beginningtobabbleapp.models.Notification;

import java.util.List;

import rx.Subscription;

public class UserInfoActivity extends AppCompatActivity implements UserInfoViewInterface {
    EditText birthDayEditText;
   EditText zipCodeEditText;
     EditText firstNameEditText;
    Button saveButton;
    TextView pleaseTapHereTextView;
    Boolean newUser;
    ProgressDialog mDialog;
    Subscription savePlayerSubscrtion;
    Subscription retriveNotificationsSubscrtion;
    List<Notification> awsNotifications;
    UserInfoPresenter userInfoPresenter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_user_profile);
        setUpViews();
        newUser = getIntent().getBooleanExtra("newUser",false);
        userInfoPresenter = new UserInfoPresenter(this,newUser,this);
        if (!newUser){
           userInfoPresenter.loadPlayerFromSharedPref();
        }
    }

    private void setUpViews(){
        birthDayEditText = (EditText) findViewById(R.id.userProfileFragmentBirtdayEditText);
        zipCodeEditText = (EditText) findViewById(R.id.userProfileFragmentZipCodeEditText);
        firstNameEditText = (EditText) findViewById(R.id.userProfileFragmentFirstNameEditText);
        saveButton = (Button)findViewById(R.id.userProfileFragmentSaveButton);
        pleaseTapHereTextView = (TextView)findViewById(R.id.pleaseTapHereTextView);
        setupOnClickListners();
    }

    private void setupOnClickListners(){
        pleaseTapHereTextView.setOnClickListener(view -> whyActivityIntent());
        saveButton.setOnClickListener(view -> saveButtonClicked());

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        userInfoPresenter.onDestory();
    }

    private void whyActivityIntent() {
        Intent whyIntent = new Intent(this, WhyActivity.class);
        startActivity(whyIntent);
    }


    private void saveButtonClicked(){
        userInfoPresenter.createBabblePlayer(createBabblePlayer());
        userInfoPresenter.updatePlayer();

    }


    private BabblePlayer createBabblePlayer(){
        BabblePlayer babblePlayer = new BabblePlayer();
        babblePlayer.setBabyName(firstNameEditText.getText().toString().trim());
        babblePlayer.setZipCode(Integer.parseInt(zipCodeEditText.getText().toString().trim()));
        babblePlayer.setBabyBirthday(birthDayEditText.getText().toString().trim());
        babblePlayer.setBabbleID(Settings.Secure.getString(getContentResolver(), Settings.Secure.ANDROID_ID));
        return babblePlayer;
    }



    @Override
    public void displayErrorDialog(int dialogTitle, int dialogMessage) {
        this.runOnUiThread(() -> Utility.displayAlertMessage(dialogTitle,dialogMessage,this));

    }

    @Override
    public void dismissActivity() {
        UserInfoActivity.this.finish();
    }

    @Override
    public void downloadIntent() {
        Intent downloadIntent = new Intent(UserInfoActivity.this,DownloadActivity.class);
        startActivity(downloadIntent);
    }

    @Override
    public void displayUserInfo(BabblePlayer babblePlayer) {
        birthDayEditText.setText(babblePlayer.getBabyBirthday());
        firstNameEditText.setText(babblePlayer.getBabyName());
        zipCodeEditText.setText(String.valueOf(babblePlayer.getZipCode()));
    }

    @Override
    public void displaySaveDialog() {
        this.runOnUiThread(() -> createDialog() );
    }

    private void createDialog(){
        mDialog = new ProgressDialog(UserInfoActivity.this);
        mDialog.setMessage("Saving Player Info");
        mDialog.setTitle("Beginning to Babble");
        mDialog.show();
    }

    @Override
    public void dismissSaveDialog() {
        this.runOnUiThread(() -> mDialog.dismiss());
    }
}

File: babbleapp/src/main/java/com/learning/leap/bwb/settings/userInfo/UserInfoActivity.kt
package com.learning.leap.bwb.settings.userInfo

import android.app.ProgressDialog
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.learning.leap.bwb.R
import com.learning.leap.bwb.download.DownloadActivity
import com.learning.leap.bwb.helper.LocalLoadSaveHelper
import com.learning.leap.bwb.model.BabbleUser
import com.learning.leap.bwb.settings.WhyActivity
import com.learning.leap.bwb.userInfo.UserInfoPresenter
import com.learning.leap.bwb.userInfo.UserInfoViewInterface
import com.learning.leap.bwb.utility.Constant
import com.learning.leap.bwb.utility.DynamoDBSingleton
import com.learning.leap.bwb.utility.NetworkChecker
import com.learning.leap.bwb.utility.Utility
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog
import io.realm.Realm
import kotlinx.android.synthetic.main.activity_user_info.*
import java.util.*


class UserInfoActivity : AppCompatActivity(), UserInfoViewInterface, DatePickerDialog.OnDateSetListener {
    var newUser: Boolean = false
    var mDialog: ProgressDialog? = null
    var gender: String = "Now Now"
    var date = ""
    var dialogIsShowing = false

    lateinit var userInfoPresenter: UserInfoPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_info)
        onClickListners()
        setUpDateTextView()
        setActionBarBackgroundColor()
        newUser = intent.getBooleanExtra(Constant.NEW_USER, false)
        userInfoPresenter = UserInfoPresenter(newUser,this, LocalLoadSaveHelper(this),
                NetworkChecker(this),Realm.getDefaultInstance(),
                DynamoDBSingleton.getDynamoDB(this))
        if (!newUser) {
            userInfoPresenter.loadPlayerFromSharedPref()
        }
    }

    private fun setActionBarBackgroundColor(){
        supportActionBar?.setBackgroundDrawable(ColorDrawable(ContextCompat.getColor(this,R.color.dark_green)))
        supportActionBar?.title = "User Information"
        if (!newUser) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId: Int = item.itemId
        if (itemId == android.R.id.home) {
            finish()
        }
        return true

    }

    private fun setUpDateTextView(){
        userProfileBirtdayEditText.isFocusable = false
        userProfileBirtdayEditText.isClickable = true
        val now = Calendar.getInstance()
        val dpd: DatePickerDialog = DatePickerDialog.newInstance(
                this,
                now[Calendar.YEAR],  // Initial year selection
                now[Calendar.MONTH],  // Initial month selection
                now[Calendar.DAY_OF_MONTH] // Inital day selection
        )
        dpd.accentColor = ContextCompat.getColor(this,R.color.darkestBlue)
        dpd.setOkColor(Color.WHITE)
        dpd.setCancelColor(Color.WHITE)
        dpd.setOnCancelListener {
            dialogIsShowing = false
        }
        userProfileBirtdayEditText.setOnClickListener {
            if (!dialogIsShowing){
                dialogIsShowing = true
                dpd.show(this.supportFragmentManager, null)
            }

        }
    }

    private fun onClickListners() {
        pleaseTapHereTextView.setOnClickListener { view: View? -> whyActivityIntent() }
        userProfileSaveButton.setOnClickListener { view: View? -> saveButtonClicked() }
        userMaleButton.setOnClickListener {
            maleButtonSelected()
            gender = "Male"
        }

        userFemaleButton.setOnClickListener {
            femaleButonSelected()
            gender = "Female"
        }

        userNotNowButton.setOnClickListener {
            notNowSelected()
            gender = "Not Now"
        }
    }

    private fun notNowSelected() {
        userMaleButton.backgroundTintList = greyBackgroundTint()
        userFemaleButton.backgroundTintList = greyBackgroundTint()
        userNotNowButton.backgroundTintList = darkBlueeBackgroundTint()
    }

    private fun femaleButonSelected() {
        userMaleButton.backgroundTintList = greyBackgroundTint()
        userFemaleButton.backgroundTintList = darkBlueeBackgroundTint()
        userNotNowButton.backgroundTintList = greyBackgroundTint()
    }

    private fun maleButtonSelected() {
        userMaleButton.backgroundTintList = darkBlueeBackgroundTint()
        userFemaleButton.backgroundTintList = greyBackgroundTint()
        userNotNowButton.backgroundTintList = greyBackgroundTint()
    }

    private fun greyBackgroundTint():ColorStateList?{
       return ContextCompat.getColorStateList(this, R.color.light_grey)
    }

    private fun darkBlueeBackgroundTint():ColorStateList?{
        return ContextCompat.getColorStateList(this, R.color.darkestBlue)
    }

    override fun onDestroy() {
        super.onDestroy()
        userInfoPresenter.onDestroy()
    }

    private fun whyActivityIntent() {
        val whyIntent = Intent(this, WhyActivity::class.java)
        startActivity(whyIntent)
    }

    private fun saveButtonClicked() {
        userInfoPresenter.createBabblePlayer(createBabblePlayer())
        userInfoPresenter.checkUserInput()
    }

    private fun createBabblePlayer(): BabbleUser {
            return BabbleUser(UUID.randomUUID().toString(), date,
                    userProfileFragmentFirstNameEditText.text.toString().trim { it <= ' ' },gender)

    }

    override fun displayErrorDialog(dialogTitle: Int, dialogMessage: Int) {
        runOnUiThread { Utility.displayAlertMessage(dialogTitle, dialogMessage, this) }
    }

    override fun displayUserInfo(babbleUser: BabbleUser) {
        userProfileBirtdayEditText.setText(babbleUser.babyBirthday)
        userProfileFragmentFirstNameEditText.setText(babbleUser.babyName)
        when (babbleUser.babyGender) {
            "Male" -> maleButtonSelected()
            "Female" -> femaleButonSelected()
            else -> notNowSelected()
        }
    }

    override fun dismissActivity() {
        Utility.addCustomEvent(Constant.CHANGE_PROFILE_SETTINGS, Utility.getUserID(this), null)
        finish()
    }

    override fun downloadIntent() {
        val downloadIntent = Intent(this@UserInfoActivity, DownloadActivity::class.java)
        startActivity(downloadIntent)
    }


    override fun getUserGender(): String {
        return gender
    }

    override fun displaySaveDialog() {
        runOnUiThread { createDialog() }
    }

    private fun createDialog() {
        mDialog = ProgressDialog(this@UserInfoActivity)
        mDialog!!.setMessage("Saving Player Info")
        mDialog!!.setTitle("Beginning to Babble")
        mDialog!!.show()
    }

    override fun dismissSaveDialog() {
        if (mDialog != null) {
            runOnUiThread { mDialog!!.dismiss() }
        }
    }

    override fun onDateSet(view: DatePickerDialog?, year: Int, monthOfYear: Int, dayOfMonth: Int) {
        val correctMonth = monthOfYear+1
       date = "$correctMonth/$dayOfMonth/$year"
        userProfileBirtdayEditText.setText(date)
    }
}